clientapi/routing/getevent.go

```diff
@@ -0,0 +1,178 @@
+		}
+
+		txnResp, err := federation.GetEvent(req.Context(), domain, eventID)
+		if err != nil {
+			return httputil.LogThenError(req, err)
+		}
+
+		if len(txnResp.PDUs) == 0 {
+			return util.JSONResponse{
+				Code: http.StatusNotFound,
+				JSON: jsonerror.NotFound("The event was not found or you do not have permission to read this event."),
+			}
+		}
+
+		requestedEvent = txnResp.PDUs[0]
+	} else {
+		requestedEvent = eventsResp.Events[0]
+	}
+
+	r := getEventRequest{
+		req:            req,
+		device:         device,
+		roomID:         roomID,
+		eventID:        eventID,
+		cfg:            cfg,
+		federation:     federation,
+		keyRing:        keyRing,
+		requestedEvent: requestedEvent,
+	}
+
+	stateNeeded := gomatrixserverlib.StateNeededForAuth([]gomatrixserverlib.Event{r.requestedEvent})
+	stateReq := api.QueryStateAfterEventsRequest{
+		RoomID:       r.requestedEvent.RoomID(),
+		PrevEventIDs: r.requestedEvent.PrevEventIDs(),
+		StateToFetch: stateNeeded.Tuples(),
+	}
+	var stateResp api.QueryStateAfterEventsResponse
+	if err := queryAPI.QueryStateAfterEvents(req.Context(), &stateReq, &stateResp); err != nil {
+		return httputil.LogThenError(req, err)
+	}
+
+	if !stateResp.RoomExists {
+		return util.JSONResponse{
+			Code: http.StatusNotFound,
+			JSON: jsonerror.NotFound("The event was not found or you do not have permission to read this event."),
+		}
+	}
+
+	if !stateResp.PrevEventsExist {
+		// Missing some events locally so stateResp.StateEvents will be unavailable.
+		// Do a federation query in hope of getting the state events needed.
+		return r.proceedWithMissingState()
+	}
+
+	return r.proceedWithStateEvents(stateResp.StateEvents)
+}
+
+// proceedWithMissingState tries to proceed by fetching the missing states with
+// federation.
+// Note: It's not guaranteed that the server(s) we query have the state events.
+func (r *getEventRequest) proceedWithMissingState() util.JSONResponse {
+	_, domain, err := gomatrixserverlib.SplitID('$', r.eventID)
+	if err != nil {
+		return httputil.LogThenError(r.req, err)
+	}
+
+	if domain == r.cfg.Matrix.ServerName {
+		// Don't send a federation query to self
+		return util.JSONResponse{
+			Code: http.StatusNotFound,
+			JSON: jsonerror.NotFound("The event was not found or you do not have permission to read this event."),
+		}
+	}
+
+	state, err := r.federation.LookupState(r.req.Context(), domain, r.roomID, r.eventID)
+	if err != nil {
+		return httputil.LogThenError(r.req, err)
+	}
+
+	if err := state.Check(r.req.Context(), r.keyRing); err != nil {
+		return httputil.LogThenError(r.req, err)
+	}
+
+	return r.proceedWithStateEvents(state.StateEvents)
+}
+
+func (r *getEventRequest) proceedWithStateEvents(stateEvents []gomatrixserverlib.Event) util.JSONResponse {
+	allowed := false
+	for _, stateEvent := range stateEvents {
+		if stateEvent.StateKeyEquals(r.device.UserID) {
+			membership, err := stateEvent.Membership()
+			if err == nil && membership == "join" {
+				allowed = true
```

Comment:
We can't use `QueryMembershipForUser` before we want the state at a specific event, not the current state.

But I think there's a new problem after I read the implementation of `QueryStateAfterEvents` again: If the user wants to get a `leave` event for themselves, then the state after that event will be the user is not in the room, and the request will thus be denied. So we probably need to get the state *at* the event instead of *after* the event.

Reply:
Interestingly I just tested the above edge case with Synapse and it will 403 you from retrieving the leave event since you aren't in the room...

Reply:
The spec doesn't say anything specific about this, so let's keep it simple for now and go Synapse's way.

Reply:
Oh, it seems `QueryMembershipForUser` still isn't the best because we need to ensure the event isn't before the user's join event.

Reply:
Previously I talked about the edge case if we use `QueryStateAfterEvents`. That problem should be nonexistent because `QueryStateAfterEvents` queries the state after _prevEvents_, which means the state at the _requested_ event.

Reply:
But Synapse should probably fix its 403. From the [spec](https://matrix.org/docs/spec/client_server/r0.5.0#post-matrix-client-r0-rooms-roomid-leave) for `/leave`:

> The user will still be allowed to retrieve history from the room which they were previously allowed to see.